Added parent validation for auto date histogram

jklancic opened the pull request:
Added auto date histogram validation for parent aggregation to derivative, cumulative sum, moving average, moving function and serial differencing aggregations.

elasticmachine commented:
Pinging @elastic/es-search-aggs

dliappis commented:
@polyfractal since you looked at the parent issue #35578 maybe you'd like to take a look at this PR?

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/AbstractPipelineAggregationBuilder.java:
```diff
@@ -102,6 +105,31 @@ public final PipelineAggregator create() throws IOException {
+        if (parent instanceof HistogramAggregatorFactory) {
+            HistogramAggregatorFactory histoParent = (HistogramAggregatorFactory) parent;
+            if (histoParent.minDocCount() != 0) {
+                throw new IllegalStateException("parent histogram of derivative aggregation [" + name
```
Since this is shared by a bunch of aggs now, I think we need to tweak the exception message to use the `type` variable:

`"parent histogram of " + type + " aggregation..."`

Same for the other exception below.

I'm not positive, but this may affect some of the unit tests elsewhere (we sometimes check exception messages).  Just FYI some of the tests may need slight tweaking if the exception message changes.

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/SerialDiffPipelineAggregationBuilder.java:
```diff
@@ -134,6 +138,12 @@ protected DocValueFormat formatter() {
     }
+    
+    @Override
+    public void doValidate(AggregatorFactory<?> parent, Collection<AggregationBuilder> aggFactories,
```
++ Thanks for adding this

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/AbstractPipelineAggregationBuilder.java:
```diff
@@ -102,6 +105,31 @@ public final PipelineAggregator create() throws IOException {
+            }
+        } else {
+            throw new IllegalStateException("derivative aggregation [" + name
+                    + "] must have a histogram, date_histogram or auto_datehistogram as parent");
```
Tiny nit: should be `auto_date_histogram` rather than `auto_datehistogram`

jklancic commented:
@polyfractal thanks for looking into this. But I still have a question. Like I wrote in the [initial bug issue](https://github.com/elastic/elasticsearch/issues/35578), the auto date histogram does not have a state like `min_doc_count`. I would need to add it to the builder class as well as to the factory class. That state is completely missed. That is the reason I thought that not the same condition can be applied for validation. BUT I can add that and all the changes you suggested up.

polyfractal commented:
@jklancic No problem.  I think we can safely ignore the `min_doc_count` validation for the `auto_date_histogram`.  The agg doesn't have a min doc count property at all (it's not an option to set on the aggregation), so we don't need to verify it.

It's not an option because that particular histogram works very differently from the others, and so the edge-case we were protecting against in the validation isn't needed.

jklancic commented:
All done @polyfractal > I addressed your comments (and run all the tests locally), please let me know if there is anything else that needs to be changed. Thanks!

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/AbstractPipelineAggregationBuilder.java:
```diff
@@ -102,6 +104,26 @@ public final PipelineAggregator create() throws IOException {
+            if (histoParent.minDocCount() != 0) {
+                throw new IllegalStateException("parent histogram of " + type + " aggregation [" + name + "] must have min_doc_count of 0");
+            }
+        } else {
```
Ah sorry, this is my fault I wasn't clear enough.  We'll still need to make sure the parent agg is `HistogramAggregatorFactory`, `DateHistogramAggregatorFactory` or `AutoDateHistogramAggregatorFactory`.  It's just that the `AutoDateHistogramAggregatorFactory` case won't have any additional validation like checking min_doc_count.

So it'll just be an empty if clause.  We still need to make sure it's one of the three classes though, otherwise the exception below will throw like the original issue.

jklancic commented:
@polyfractal hope that should resolve the issue. Sorry for removing it - I thought it is strange to remove this, since the bug issue was filled to resolve this exact thing. I took the liberty and changed the logic a bit to reflect what you told me. I can still revert and just add an empty if statement, but I feel that the newest version better emphasizes the validation that takes place.

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/pipeline/AbstractPipelineAggregationBuilder.java:
```diff
@@ -102,6 +105,28 @@ public final PipelineAggregator create() throws IOException {
+     * Validates pipeline aggregations that need sequentially ordered data.
+     */
+    public static void validateSequentiallyOrderedParentAggs(AggregatorFactory<?> parent, String type, String name) {
+        if (!(parent instanceof HistogramAggregatorFactory || parent instanceof DateHistogramAggregatorFactory
```
++ This arrangement is cleaner than an empty `if`, thanks!

One super tiny style nit:  could you change it from a `!(...)` to an `(...) == false`?  We try not to use the negation operator because it's easy to misread, whereas the more verbose `== false` is hard to miss.

MirAkram commented:
mj1265849@gmail.com
در تاریخ ۲۰ نوامبر ۲۰۱۸ ۶:۵۱ بعدازظهر، "Zachary Tong" <
notifications@github.com> نوشت:

> *@polyfractal* requested changes on this pull request.
>
> Thanks! Left a tiny comment about style.
>
> Would you be able/willing to add some tests? It'll be a little more
> involved than the refactor itself, so if you're too busy or don't want to,
> I'd be happy to help out.
>
> But if you're willing, I think we should add some unit tests to the
> relevant pipeline aggs (derivative, cusum, movavg, movfn, serial diff) to
> test that these validations work and to help guard against accidental
> breaks in the future.
>
> The unit tests can be found here
> <https://github.com/elastic/elasticsearch/tree/master/server/src/test/java/org/elasticsearch/search/aggregations/pipeline>
> and are the ones that don't have "IT" in the name ("IT" stands for
> "integration test" and those test against actual nodes, overkill for this
> case). The StatsBucket has a validation test that you can use as a base
> <https://github.com/elastic/elasticsearch/blob/master/server/src/test/java/org/elasticsearch/search/aggregations/pipeline/StatsBucketTests.java#L39>.
> The validations will be different (exceptions if pointing to non-histo
> aggs, no exception for histo aggs, etc) but that'll provide the boilerplate.
>
> Let me know if you're up for adding the tests, or if you have any
> questions.
>
> Thanks again for working on this, really appreciate it!
> ------------------------------
>
> In server/src/main/java/org/elasticsearch/search/aggregations/pipeline/
> AbstractPipelineAggregationBuilder.java
> <https://github.com/elastic/elasticsearch/pull/35670#discussion_r235005671>
> :
>
> > @@ -102,6 +105,28 @@ public final PipelineAggregator create() throws IOException {
>      public void doValidate(AggregatorFactory<?> parent, Collection<AggregationBuilder> factories,
>              Collection<PipelineAggregationBuilder> pipelineAggregatorFactories) {
>      }
> +
> +    /**
> +     * Validates pipeline aggregations that need sequentially ordered data.
> +     */
> +    public static void validateSequentiallyOrderedParentAggs(AggregatorFactory<?> parent, String type, String name) {
> +        if (!(parent instanceof HistogramAggregatorFactory || parent instanceof DateHistogramAggregatorFactory
>
> ++ This arrangement is cleaner than an empty if, thanks!
>
> One super tiny style nit: could you change it from a !(...) to an (...)
> == false? We try not to use the negation operator because it's easy to
> misread, whereas the more verbose == false is hard to miss.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/35670#pullrequestreview-176769822>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AqW2IwH8cfCvtIp3RZ17T4mcVV5UwY6dks5uxAjngaJpZM4YnGDj>
> .
>

jklancic commented:
@polyfractal I added the change to the validation as well as unit tests. Let me know if you want any changes regarding the unit test. I also run all test and they completed successfully.

polyfractal commented:
Thanks @jklancic!  Looks good!  (sorry for the delay, yesterday was a zoo here)

I have one last request and then I think this is good to go.  Could you add to the tests a section where we test one of `[histogram, date_histogram, auto_date_histogram]` to make sure it works as expected?

E.g. pick one of the three randomly, set the pipeline to point at the chosen agg, validate and make sure no exceptions are thrown.

Thanks!  I appreciate your patience on this one :)

jklancic commented:
@polyfractal all done now. But I really hope this was the last change :)

polyfractal commented:
:)  Tests and the like are always more work than the original fix in my experience.  Thanks for sticking through all the changes!

@elasticmachine test this please

polyfractal commented:
@jklancic Looks like there are some checkstyle violations, based on the CI tests.  Probably unused imports or line-lengths.

If you run `./gradlew precommit` that will run the checkstyle locally so you can see what broke.

(If at any point you get tired of working on this lemme know, I can push it over the finish line and you'll still be the merge-commit author)

jklancic commented:
@polyfractal I think I got all the formatting issues. I ran all the tests again as well, to make sure they still run successfully.

polyfractal commented:
@elasticmachine test this please

polyfractal commented:
@elasticmachine run the gradle build tests 1

polyfractal commented:
@elasticmachine test this please

@jklancic There was a failing test due to something unrelated, and was fixed in master.  I went ahead and merged master so that we could kick off the CI build again.  I hope that was ok. :)

polyfractal commented:
@elasticmachine run the gradle build tests 2

polyfractal commented:
@elasticmachine test this please

jklancic commented:
@polyfractal can we merge this now :) Seems like everything is green

polyfractal commented:
Ah my fault, slipped through the crack.  The distributions test failed (unrelated), so I merged master and pushed.  I also made a small tweak, moving the random parent agg helper out of `ESTestCase` and into a pipeline helper class.  I missed that earlier.  

It's not a big deal, it's just `ESTestCase` is shared by most of the tests so we try to keep it as "clean" as possible.  Made sense to relocate the random parent helper to something closer to the pipeline tests.

Will keep a close eye on the tests and get this merged ASAP.  Thanks for your patience, and sorry for letting it drop!

@elasticmachine test this please

polyfractal commented:
Merged!  Thanks for sticking with it @jklancic, and apologies it took so long to get merged!

I'm backporting to 6.x right now, will push a commit after I tidy up a few merge conflicts.
